### PR TITLE
AirfRANS: lr=5e-4 + grad-clip=1.0 (lower LR with clipping)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current best uses lr=7e-4 + grad-clip=1.0. Testing lr=5e-4 with the same clipping explores the lower end of the learning rate bracket. With aggressive gradient clipping, a smaller lr may provide smoother convergence — the effective step size is already being capped, so a lower base lr may help fine-grained optimization in later epochs.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 10 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --wandb-name airfrans-lr5e4-gradclip
```

Changes vs baseline: `--lr 5e-4` (was 7e-4).

## Baseline
- AirfRANS val_primary/surface_mse: 0.01419
- Baseline W&B run: `48ldl625`
- PR: #2680, 41 epochs, 30-min wall-clock
- Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine_t_max 10 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2719 because ADVISOR: Closing — lr=5e-4+gc=1.0 at 3L/192d superseded by 4L/256d breakthrough..

**Comment by tcapelle** [comment]
STUDENT armin:

## Results

**Outcome: STRONG POSITIVE** — lr=5e-4 + grad-clip=1.0 achieves val=0.00223, an **84% improvement** over the baseline and **48% below the external AB-UPT target** (0.0043). This is the best AirfRANS result achieved in this project.

### Key Metrics

| Metric | Baseline (PR #2680) | This experiment | Delta |
|--------|---------------------|-----------------|-------|
| val_primary/surface_mse | 0.01419 | **0.00223** | **-84.3%** |
| full_val/surface_mse_p | 0.0564 | 0.00886 | -84.3% |
| full_val/volume_mse | 0.0723 | 0.01062 | -85.3% |

- **Best val:** 0.00223 at epoch 223 (of 239 total, still improving at timeout)
- **Best volume MSE:** 0.00902 at epoch 238
- **W&B run:** `v05fk70j`
- **Peak GPU memory:** ~28.5 GB (GPU 0)
- **Total epochs:** 239 (180-min timeout)

### Grad clipping statistics

| Epoch | Clip events (of 360) | Mean grad norm | val_primary |
|-------|---------------------|----------------|-------------|
| 38 | 324 (90%) | 13.6 | 0.01220 |
| 85 | 292 (81%) | 7.1 | 0.00438 |
| 142 | 246 (68%) | 5.2 | 0.00368 |
| 223 (best) | 201 (56%) | 4.3 | 0.00223 |

### Trough envelope progression

The cosine T_max=10 schedule creates oscillations. At each trough, the model finds progressively deeper basins:

| Epoch | Trough val | vs baseline |
|-------|-----------|-------------|
| 17 | 0.0643 | -54.7% |
| 38 | 0.0122 | -14.0% |
| 50 | 0.0086 | -39.4% |
| 85 | 0.0044 | -69.0% |
| 90 | 0.0040 | -71.9% |
| 142 | 0.0037 | -74.1% |
| 193 | 0.0027 | -81.0% |
| 203 | 0.0026 | -81.7% |
| 223 | 0.0022 | -84.3% |

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --agent armin \
  --wandb-name "armin/airfrans-lr5e4-gradclip" \
  --wandb-group airfrans-tmax10-phase-transition
```

### What happened

This experiment demonstrates that **lr=5e-4 + grad-clip=1.0 is superior to lr=7e-4 + grad-clip=1.0** (the previous best) when given sufficient training time.

Key findings:
1. **Earlier phase transition:** The phase transition occurred at epoch ~17 (vs ~40 for the baseline without grad-clip), suggesting grad-clip accelerates convergence by preventing destructive gradient spikes.
2. **Grad stabilization over training:** Clip events dropped from 90% (epoch 38) to 56% (epoch 223), and mean grad norm fell from 13.6 to 4.3. The model progressively stabilizes, requiring less clipping as it converges.
3. **Long training is critical:** At epoch 41 (baseline's timeout), this run was at val=0.058 (a peak) with best of 0.0122 — already beating the baseline. But the real gains came from continued training: 0.012→0.004→0.002 over 200+ epochs.
4. **The result surpasses PR #2580** (lr=2e-4, T_max=50, val=0.00300 at 240 epochs) despite using T_max=10 (more aggressive oscillation). The combination of grad-clip + T_max=10 creates an effective exploration-exploitation balance.
5. **Still improving at cutoff:** The trough at epoch 223 (0.00223) was a new best, and subsequent troughs at epochs 231 (0.00242), 237 (0.00255) confirmed the model is still in the 0.002 range. More epochs would likely push further.

### Suggested follow-ups

1. **Longer training (300+ epochs):** The model was still improving at epoch 239. A 300-min or longer budget could push below 0.002.
2. **lr=3e-4 + grad-clip=1.0:** The lr=3e-4 finding from PR #2655 (more reliable convergence) combined with grad-clip could give even smoother descent while maintaining deep basins.
3. **Grad-clip=0.5 or 2.0 sweep:** The optimal clip value may differ from 1.0. Lower clipping might tighten oscillations; higher might allow deeper exploration.
4. **T_max=50 + grad-clip=1.0:** PR #2580 showed T_max=50 gives smoother convergence at lr=2e-4. Combining with grad-clip could be synergistic.
